### PR TITLE
Make the `LanguageServerProtocolJSONRPC` module build with strict concurrency enabled

### PR DESCRIPTION
**Pull request by ahoppen**
Also use this opportunity to overhaul the entire connection handling. 

Probably best reviewed commit by commit.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift [hunk 1]
```diff
@@ -21,53 +21,91 @@ import struct CDispatch.dispatch_fd_t
...
+public final class JSONRPCConnection: Connection {
 
   /// A name of the endpoint for this connection, used for logging, e.g. `clangd`.
-  let name: String
+  private let name: String
 
-  var receiveHandler: MessageHandler? = nil
+  /// The message handler that handles requests and notifications sent through this connection.
+  ///
+  /// Access to this must be be guaranteed to be sequential to avoid data races. Currently, all access are
```
Comment (bnbarham) [inline Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift:32]: `serial` or `synchronized` would make more sense to me than `sequential` here. Same with the other places

**Review by bnbarham — APPROVED** [review]

File: Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift [hunk 2]
```diff
@@ -22,52 +22,69 @@ import struct CDispatch.dispatch_fd_t
...
+  ///  - `init`: Can trivially only be called once
+  ///  - `start`: Is required to be call in the same sequential code region as the initializer, so
+  ///    `JSONRPCConnection` can't have escaped to other isolation domains yet.
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift:35]: Good idea with calling `start` on `queue` as well.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Pull request merged**